core/src/main/java/org/apache/druid/timeline/DataSegment.java

```diff
@@ -76,9 +76,22 @@
     @Inject(optional = true) @PruneLoadSpec boolean pruneLoadSpec = false;
   }
 
+  public static DataSegment intern(DataSegment dataSegment, boolean updateLoadSpec)
+  {
+    DataSegment result = DATA_SEGMENT_INTERNER.intern(dataSegment);
+    if (updateLoadSpec) {
+      result.dimensions = dataSegment.dimensions;
```

Comment:
I think `DataSegment` shouldn't become mutable. However, it would be nice if you would solve this problem in this PR: #6358.